src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Internal/MatcherContext.cs

```diff
@@ -121,129 +144,255 @@ private void Match(DirectoryInfoBase directory, string? parentRelativePath)
+
+                foreach (IPatternContext context in _excludePatternContexts)
+                {
+                    context.PushDirectory(directory);
+                }
             }
         }
 
-        internal static string CombinePath(string? left, string right)
+        private void MatchOrdered(DirectoryInfoBase directory, string? parentRelativePath)
         {
-            if (string.IsNullOrEmpty(left))
+            // Request all the including and excluding patterns to push current directory onto their status stack.
+            PushDirectory(directory);
+            Declare();
+
+            var entities = new List<FileSystemInfoBase?>();
+            if (_declaredWildcardPathSegment || _declaredLiteralFileSegments.Count != 0)
             {
-                return right;
+                entities.AddRange(directory.EnumerateFileSystemInfos());
             }
             else
             {
-                return $"{left}/{right}";
+                IEnumerable<FileSystemInfoBase> candidates = directory.EnumerateFileSystemInfos();
+                foreach (FileSystemInfoBase candidate in candidates)
+                {
+                    if (candidate is DirectoryInfoBase &&
+                        _declaredLiteralFolderSegmentInString.Contains(candidate.Name))
+                    {
+                        entities.Add(candidate);
+                    }
+                }
             }
-        }
 
-        // Used to adapt Test(DirectoryInfoBase) for the below overload
-        private bool MatchPatternContexts<TFileInfoBase>(TFileInfoBase fileinfo, Func<IPatternContext, TFileInfoBase, bool> test)
-        {
-            return MatchPatternContexts(
-                fileinfo,
-                (ctx, file) =>
+            if (_declaredParentPathSegment)
+            {
+                entities.Add(directory.GetDirectory(".."));
+            }
+
+            var subDirectories = new List<DirectoryInfoBase>();
+
+            foreach (var entity in entities)
+            {
+                if (entity is FileInfoBase fileInfo)
                 {
-                    if (test(ctx, file))
+                    PatternTestResult result = MatchWithOrder(fileInfo);
+                    if (result.IsSuccessful)
                     {
-                        return PatternTestResult.Success(stem: string.Empty);
+                        _files.Add(new FilePatternMatch(
+                            path: CombinePath(parentRelativePath, fileInfo.Name),
+                            stem: result.Stem));
                     }
-                    else
+
+                    continue;
+                }
+
+                if (entity is DirectoryInfoBase dirInfo)
+                {
+                    bool result = false;
+                    foreach (var (isInclude, context) in _orderedContexts)
                     {
-                        return PatternTestResult.Failed;
+                        if (context.Test(dirInfo))
```

Comment:
Say we have the following patterns:
- include A/*
- exclude A/B/*
- include C/*

If we have a directory, C, that we are checking, then we can: fail matching against the 1st pattern, ignore the 2nd pattern, and succeed matching against the last pattern. The current implementation doesn't ignore the 2nd pattern. You should be able to change the logic here to `if ((result != isInclude) && context.Test(dirInfo))` to add this short-circuiting. Same change can be applied in `MatchWithOrder`.

Reply:
It's not a correctness issue, just a perf improvement.